Add GEP-3155: Complete Backend mTLS Configuration (#3180)
kubernetes-sigs/gateway-api: geps/gep-3155/index.md

Reviewer: Just a reminder, you'll need to also update `mkdocs.yml` so that it shows in https://gateway-api.sigs.k8s.io/geps/overview/ as well as  `geps/gep-3155/metadata.yaml`

Author: Thanks, done

Reviewer 2: Hello @mkosieradzki , if I understood the last iterations of GEP correctly, we are proposing that the BackendTLSPolicy spec target reference will be only gateway and not service ? If that is true , how would the scenarios where there are mix of services which do not need MTLS coexist in the same Gateway -> Route hierarchy.

* Status: Implementable

## TLDR

This GEP aims to complete the configuration required for Backend mutual TLS in Gateway
API. This includes the following new capabilities:

1. Configuration for the client certificate Gateways should use when connecting
   to Backends
1. Ability to specify SANs on BackendTLSPolicy
1. Add TLS options to BackendTLSPolicy to mirror TLS config on Gateways

## Goals

* Add sufficient configuration that basic mutual TLS is possible between Gateways and
  Backends
* Enable the optional use of SPIFFE for Backend mutual TLS

## Non-Goals

* Define how automatic mTLS should be implemented with Gateway API

## Introduction

This is a wide ranging GEP intending to cover three additions to the API that all
have a shared goal - enabling backend mutual TLS with Gateway API. Although this
specific GEP focuses on manual configuration across the board, the hope is that
it will also enable higher level automation to simplify this process for users.

## API

### Client Certs on Gateways

A key requirement of mutual TLS is that the Gateway can provide a client cert to the
backend. This adds that configuration to both Gateway and Service (via 
BackendTLSPolicy).

> #### Gateway-level (Core support)

Reviewer 3: This is missing an API example - seems to be suggesting the addition of a new field under the Gateway `spec` stanza https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.Gateway?

Reviewer 4: Yes, I think this should have a quick sketch of what we need to add to Gateway.

Specifying credentials at the gateway level is the default operation mode, where all

> backends will be presented with a single gateway certificate.

Reviewer 3: Re `all` backends - should the client cert be presented even when no BackendTLSPolicy targeting the Service backend is present?

Reviewer 3: > Ideally Gateway owners could provide a set of CAs they trust for certs from backends at the Gateway level

I think this makes sense and could potentially address the non-mutual end-to-end TLS functionality @jaishals is describing in https://github.com/kubernetes-sigs/gateway-api/pull/3180#discussion_r1697382336, where the presence of BackendTLSPolicy additional validation is specifically to enforce _mutual_ TLS? Would any additional signal like `mode: mutual` be needed somewhere to clarify this?

### SANs on BackendTLSPolicy

> This change enables the certificate to have a different identity than the SNI

Reviewer 3: I think this is intended to be applicable to the "server"/destination cert, but can you please specify more precisely which cert is affected here?

(both are currently tied to the hostname field). This is particularly useful
when using SPIFFE, which relies on URI Subject Names which are not valid SNIs 
as per https://www.rfc-editor.org/rfc/rfc6066.html#section-3.

In such case either connection properties or an arbitrary SNI, like cluster-local 
service name could be used for certificate selection, while the identity validation
will be done based on SubjectAltNames field.
When specified, at least one of certificate's Subject Alternate Names MUST match at least one of the specified SubjectAltNames.

**1. Add a new `SubjectAltNames` field to `BackendTLSPolicyValidation`**

```go
type BackendTLSPolicyValidation struct {
  // SubjectAltNames contains one or more Subject Alternative Names.
  // When specified, at least one of certificate's Subject Alternate Names MUST 

>   // match at least one of the specified SubjectAltNames.

Reviewer 3: What happens if it doesn't? Suggested `status` condition on the policy indicated it's invalid?

Reviewer 5: Apparently we haven't defined _how_ a request should be rejected here:

https://github.com/kubernetes-sigs/gateway-api/blob/c94d4010a1524e86ffd07bc5ccedde340dba5843/apis/v1alpha3/backendtlspolicy_types.go#L128-L130

At a minimum I think we can agree that the request should be rejected, ideally we can be more specific here for the sake of writing conformance tests and overall portability.

As far as populating status, I think that most controllers are separated from the dataplane and would not be able to translate runtime TLS validation errors to config time (k8s) status.

>   // When specified, at least one of certificate's Subject Alternate Names MUST 
>   // match at least one of the specified SubjectAltNames.

Reviewer 4: Similarly to above, I think this should be:
```suggestion
  // When specified, the certificate served by the backend MUST have at a Subject Alternate Name
  // that matches one of the specified SubjectAltNames.
```

Just so it's _extremely_ clear what certificate this is talking about.

been helpful for implementation-specific TLS configurations, or simply features
that have not made it to the core API yet. It would be similarly useful to have
an identical field on BackendTLSPolicy.

Examples:
- configuration options for vendor-specific mTLS automation
- restrictions on the minimum supported TLS version or supported cipher suites

```go
type BackendTLSPolicySpec struct {
  // Options are a list of key/value pairs to enable extended TLS
  // configuration for each implementation. For example, configuring the
  // minimum TLS version or supported cipher suites.
  //
  // A set of common keys MAY be defined by the API in the future. To avoid
  // any ambiguity, implementation-specific definitions MUST use
  // domain-prefixed names, such as `example.com/my-custom-option`.
  // Un-prefixed names are reserved for key names defined by Gateway API.
  //
  // Support: Implementation-specific
  //
  // +optional
  // +kubebuilder:validation:MaxProperties=16
  Options map[AnnotationKey]AnnotationValue `json:"options,omitempty"`
}
```

## Conformance Details

Conformance tests will be written to ensure the following:

> 1. When SubjectAltNames are specified in BackendTLSPolicy:
>   - The hostname field is still used as SNI, if specified

Reviewer 5: Our docs gen requires a newline before a list starts (https://deploy-preview-3180--kubernetes-sigs-gateway-api.netlify.app/geps/gep-3155/#conformance-details)

```suggestion
1. When SubjectAltNames are specified in BackendTLSPolicy:

  - The hostname field is still used as SNI, if specified
```

  - A certificate with at least one matching SubjectAltName is accepted
  - A certificate without a matching SubjectAltName is rejected

> 2. When a Client Certificate is specified on a Gateway:
>   - It is applied to all services.

Reviewer 5: ```suggestion
2. When a Client Certificate is specified on a Gateway:

  - It is applied to all services.
```

> # GEP-3155: Complete Backend mTLS Configuration

Reviewer 4: I know and _most_ people reading this also know that "mTLS" here means "mutual TLS between parties", _without_ the additional Service Mesh concepts that make "mTLS" mean "mutual TLS between services with automatically-managed, rapidly rotated certificates".

But some people, when they read "mTLS", read the latter, not the former.

Could we please change this to "mutual TLS" throughout the document, and put a caveat at the top that "mutual TLS" in this document is used solely to refer to the two-way handshake between Gateway and backend, and _no other mutual TLS use cases_?

@mkosieradzki, I know this may seem pedantic and stupid, but in almost every conversation we've had about client certificate usage, _someone_ has been confused by the use of "mTLS". I just want to avoid having the same issue here as well.

Author: Thank you, I was not aware of that. I will update the doc accordingly to avoid confusion with managed TLS or managed mutual TLS.

  //
  // This setting can be overriden on the service level by use of BackendTLSPolicy.
  ClientCertificateRef SecretObjectReference `json:"clientCertificateRef,omitempty"`
}
```

**2. Add a new `Client` field at the top level of BackendTLSPolicy**

```go
type BackendTLSPolicySpec {
  // Client specifies overriden TLS settings for the given service.
  //
  // If Client is omitted, backend mTLS settings inherited from the given gateway are
  // used.
  Client TLSClientSettings `json:"client"`
}

type TLSClientSettings struct {
  // CertificateRef is a reference to an object that contains a Client
  // Certificate.
  //
  // References to a resource in different namespace are invalid UNLESS there
  // is a ReferenceGrant in the target namespace that allows the certificate
  // to be attached. If a ReferenceGrant does not allow this reference, the
  // "ResolvedRefs" condition MUST be set to False for this listener with the
  // "RefNotPermitted" reason.
  //
  // CertificateRef can reference to standard Kubernetes resources, i.e.
  // Secret, or implementation-specific custom resources.
  //
  // If CertificateRef is not specified, no client certificate is presented when

>   // connecting to the service.

Reviewer 3: This doesn't quite seem accurate, as it seems a client certificate could be presented still if GatewayBackendTLS were configured?

I'm also not quite clear on the use case for allowing this to be specified in two different places, could you elaborate on that in the GEP please?

Author: The idea here is to allow overriding the default from GatewayBackendTLS with no certificate.

This is to support some compatibility corner cases, for example around a pre-packaged software.

Let's imagine a prepackaged software that always requests a certificate, but if customer refuses, it still proceeds with connection but falls back to an alternative authentication method, e.g. JWT. But for some reason if it observes an unexpected certificate it rejects the connection.

If we always presented the certificate in such case, it would be impossible to use this prepackaged software correctly, without creating a dedicated gateway with TLS disabled.

Having said that, if there is a consensus that this corner case is too far fetched, I am happy to drop it.

Reviewer 3: Expecting an undefined or explicit `nil` here to override a configured value feels awkward and potentially confusing. It feels like the right way to handle this need would be to somehow scope GatewayBackendTLS better to only be applied for some backend connections?

Author: Agreed, replaced it with an Enum

  CertificateRef SecretObjectReference `json:"certificateRef,omitempty"`
}
```

#### Limitations

Configuring client certificate on the service level may result in multiple Gateways
from different vendors sharing the same identity when connecting to a single service.

The problem could be partially mitigated by allowing shared policies to selectively 

> target specific Gateway instances, GatewayClasses or gateway namespaces.

Reviewer 4: This limitation is why we did not include client certificate configuration in the initial version of BackendTLSPolicy. Similarly to @mikemorris' comment above, I'd like some use cases that illustrate the need for this, as I think that it's far more likely to be used incorrectly than only having the TLS config for a Gateway present on that Gateway.

To be honest, the BackendTLSPolicy is supposed to be owned by the same party as the Service, and I question if having the party owning the server side of the transaction also specify the client keypair to be used is really fulfilling the spirit of what mutual TLS is supposed to be about.

To put this another way - if you lock your front door, but leave the key tied with a string to the door, are you increasing or decreasing security? Part of the point of adding the second factor of client authentication to a mutual TLS transaction is to ensure that there's some checks and balances in who gets a valid client certificate, and this feels like it bypasses that.

Author: Added the use cases, added the write-up on security of this solution, can you PTAL at the updated doc?

### SANs on BackendTLSPolicy

This change enables the certificate to have a different identity than the SNI
(both are currently tied to the hostname field). This is particularly useful
when using SPIFFE, which relies on URI Subject Names which are not valid SNIs 
as per https://www.rfc-editor.org/rfc/rfc6066.html#section-3.

In such case either connection properties or an arbitrary SNI, like cluster-local 
service name could be used for certificate selection, while the identity validation

> will be done based on SubjectAltNames field.

Reviewer 4: Again, I think we may need some examples or use cases here to illustrate how this would work, as I _think_ I get it, but am not sure.

Reviewer 4: Maybe I should be clearer here. Currently, what this field is actually used for is not defined at all. Being able to specify more SANs is great, but as an implementer, what do I do with this field?

From the below struct docs, it seems like the answer is maybe "The serving certificate received when the Gateway implementation makes a client TLS connection to the backend must have at least one of the specified SANs set".

This is _critical_, because if we go ahead with also adding client certificates to this object, we now have both client and serving certificates on the same object. So it's really, really important that we leave no wiggle room here in the godoc.

Reviewer 5: +1, good points @youngnick. I think the godoc below describes the behavior fairly clearly, maybe it would be helpful to pull that up to this paragraph as well?

> When specified, at least one of certificate's Subject Alternate Names MUST match at least one of the specified SubjectAltNames.

Author: Added this sentence, does it look better now?

Reviewer 4: Sorry, but I think we really need the mention that it's the _serving_ cert that must be matched against.

So I think that sentence should be:

`When specified, the certificate served from the backend MUST have at least one Subject Alternate Name matching one of the specified SubjectAltNames`.

**1. Add a new `SubjectAltNames` field to `BackendTLSPolicyValidation`**

```go
type BackendTLSPolicyValidation struct {
  // SubjectAltNames contains one or more Subject Alternative Names.
  // When specified, at least one of certificate's Subject Alternate Names MUST 
  // match at least one of the specified SubjectAltNames.
  // +kubebuilder:validation:MaxItems=5
  SubjectAltNames []SubjectAltName `json:"subjectAltNames,omitempty"`
}

type SubjectAltName struct {
  // Hostname contains Subject Alternative Name specified in DNS name format. It is

>   // mutually exclusive with Hostname. At least one of the fields must be specified.

Reviewer 3: ```suggestion
  // mutually exclusive with URI. At least one of the fields must be specified.
```

> type SubjectAltName struct {
>   // Hostname contains Subject Alternative Name specified in DNS name format. It is
>   // mutually exclusive with Hostname. At least one of the fields must be specified.
>   Hostname v1.PreciseHostname `json:"hostname"`
> 
>   // URI contains Subject Alternative Name specified in URI format. It is mutually
>   // exclusive with Hostname. At least one of the fields must be specified.
>   URI string `json:"uri,omitempty"`
> }

Reviewer 4: I think that this struct may be better handled as either a type/value struct, as follows:
```suggestion
type SubjectAltName struct {
  // Type chooses the type of SAN stored in this struct. Valid values are either `hostname`
  // or `uri`.
  Type string `json:"type"`
  
  // Value holds the value of the SAN for the type.
  Value string `json:"value"`
}
```

or as a type-selected union, like this:

```suggestion
type SubjectAltName struct {
  // Type chooses the type of SAN stored in this struct. Valid values are either `hostname`
  // or `uri`. The corresponding field must be filled out based on the type (the
  // Hostname or URI fields respectively). If a field not corresponding to the Type is
  // nonzero, it will be ignored.
  Type string `json:"type"`
  
  // Hostname contains Subject Alternative Name specified in DNS name format. It is
  // mutually exclusive with Hostname. At least one of the fields must be specified.
  Hostname v1.PreciseHostname `json:"hostname"`

  // URI contains Subject Alternative Name specified in URI format. It is mutually
  // exclusive with Hostname. At least one of the fields must be specified.
  URI string `json:"uri,omitempty"`
}
```

The first has the advantage of brevity, at the cost of some type-safety and the associated validation on ingestion - it would need more checking in-code instead. The second keeps the type-safety, and works the same as many other type-selected union fields in Kubernetes.

I probably slightly favor the second.

Author: Thanks, I will update to type-selected union.

```

**2. Modify Spec for `BackendTLSPolicyValidation` `Hostname`**

Before:
```go
	// 2. Hostname MUST be used for authentication and MUST match the certificate
	//    served by the matching backend.
```

After:
```go
	// 2. Only when SubjectAltNames is not specified, Hostname MUST be used for 
    //    authentication and MUST match the certificate served by the matching
    //    backend.

>     // 3. If HostName is unspecified, SNI header will be omitted.

Reviewer 4: If there is no SNI, how can the server choose what certificate to serve? This implies but does not state that the server must only have a single serving certificate to choose from. If that's an assumption, I guess that's okay, but we need to be clear about the contract here.

Maybe we need to add something to some docstrings somewhere to say something like "BackendTLSPolicy expects that backends covered by it SHOULD have only one serving keypair to choose from, as the SNI field may not always be set" or something.

Author: I think that the common convention is to provide the "default" certificate in such case.

I think I would rather use language that the Hostname SHOULD be sad - otherwise server will have hard time selecting the correct certificate, but I wanted to allow unset for compatibility corner cases.

But again, if there is a consensus we don't need to support this case -> I am more than happy to drop this from the spec.

Author: Agreed, this seems to be pretty niche, let's remove it.

```

### Allow per-service mTLS settings BackendTLSPolicy

Gateway level TLS configuration already includes an `options` field. This has
been helpful for implementation-specific TLS configurations, or simply features
that have not made it to the core API yet. It would be similarly useful to have
an identical field on BackendTLSPolicy.

```go
type BackendTLSPolicySpec struct {
	// Options are a list of key/value pairs to enable extended TLS
	// configuration for each implementation. For example, configuring the
	// minimum TLS version or supported cipher suites.
	//
	// A set of common keys MAY be defined by the API in the future. To avoid
	// any ambiguity, implementation-specific definitions MUST use
	// domain-prefixed names, such as `example.com/my-custom-option`.
	// Un-prefixed names are reserved for key names defined by Gateway API.
	//
	// Support: Implementation-specific
	//
	// +optional
	// +kubebuilder:validation:MaxProperties=16
	Options map[AnnotationKey]AnnotationValue `json:"options,omitempty"`
}

> ```

Reviewer 4: I understand that these escape hatches are handy, but as a maintainer I'm concerned that they have two effects:

* They promote the creation of non-portable configuration, which is against the purpose of this API
* They encourage users to use them and not go through the (yes, sometimes arduous) process of giving feedback and building features into upstream.

The fact that you're saying this is useful and I have no idea how this field is actually being used in the field illustrates exactly what I mean. If you can get everything you need out of the spec by using a free-text escape hatch like this, why would you ever do the work to come and ask for proper typed, structured fields?

To put this another way: everytime I see a `map[string]string`, I think of Ingress annotations and the terrible situation that has ended up in. So I, personally, need to see a pretty big benefit to justify these risks. I'd like to see some concrete use cases here as well that illustrate the need for this.

Author: I believe this will be primarily used for experimental features, before they are actually merged. First two I expect are:
- constraints like min-tls-version
- configuration for proprietary / experimental managed TLS modes

backends will be presented with a single gateway certificate.

#### Service-level (Implementation-specific support)
Specifying credentials at the service level allows to provide different credentials
to connect to specific backends or completely disable client certificates. This allows
to support a wide variety of corner cases, mostly around application-compatibility, e.g.:
- given OOTB software requires clients to use self-signed, certificates issued by a
  different issuer, or certificates with a very specific SAN.
- given OOTB software uses optional certificates for end-user authentication, if the
  certificate is missing, it falls back to a different application-level authentication
  method - in such case not providing any certificate is the only way to operate

**1. Add a new `BackendTLS` field at the top level of Gateways**

```go
type GatewaySpec struct {
  // BackendTLS configures TLS settings for when this Gateway is connecting to
  // backends with TLS.
  BackendTLS GatewayBackendTLS `json:"backendTLS,omitempty"'
}

type GatewayBackendTLS struct {
  // ClientCertificateRef is a reference to an object that contains a Client

>   // Certificate.

Reviewer 4: ```suggestion
  // Certificate and the associated private key.
```

I think we need to be clear that this needs to be a full keypair, not just the certificate.

  //
  // References to a resource in different namespace are invalid UNLESS there
  // is a ReferenceGrant in the target namespace that allows the certificate
  // to be attached. If a ReferenceGrant does not allow this reference, the
  // "ResolvedRefs" condition MUST be set to False for this listener with the
  // "RefNotPermitted" reason.
  //
  // ClientCertificateRef can reference to standard Kubernetes resources, i.e.
  // Secret, or implementation-specific custom resources.
  //
  // This setting can be overriden on the service level by use of BackendTLSPolicy.
  ClientCertificateRef SecretObjectReference `json:"clientCertificateRef,omitempty"`
}
```

**2. Add a new `ClientCertificateMode` and `ClientCertificateRef` fields at the top

> level of BackendTLSPolicy**

Reviewer 4: I don't know the best way to word this, but it really feels like we need to tell people that you shouldn't use this without a good reason. Because otherwise people will look at the spec of `BackendTLSPolicy` and assume that this is the right place to configure client certificate details.

I also still think that having this config makes the client certificate used here basically worthless for actual authentication - it is essentially `curl -k` for Gateways.

As it is, I suspect that most examples we see will use this and not the Gateway level config, because this is easily within the control of the service owner, and requires no interaction with the Gateway owner.

These feelings aside, the actionable feedback here is to have some sort of callout that Client Certificate config should preferably be in the Gateway stanza, not this one. This should only be used when the Gateway option doesn't work out for some reason.

  //
  // Support: Core
  DefaultClientCertificateInheritanceMode ClientCertificateInheritanceMode = "Default"

  // Override determines the certificate specified in ClientCertificateRef field will be used.
  //
  // Support: Implementation-specific
  OverrideClientCertificateInheritanceMode ClientCertificateInheritanceMode = "Override"

  // Disable determines the certificate will be not provided to the backend, even if it is
  // configured at the gateway level and requested by the backend.
  //
  // Support: Implementation-specific
  DisableClientCertificateInheritanceMode ClientCertificateInheritanceMode = "Disable"
)

type BackendTLSPolicySpec {
  // ClientCertificateMode determines the inheritance mode for client certifacate.
  //
  // If ClientCertificateMode is unset, the value assumed is Default.
  ClientCertificateMode ClientCertificateInheritanceMode `json:"clientCertificateMode,omitempty"`

  // ClientCertificateRef is a reference to an object that contains a Client
  // Certificate.
  //
  // This value takes effect and is required only if the ClientCertificateMode is set to 
  // Override, otherwise it is ignored.
  //

>   // References to a resource in different namespace are invalid UNLESS there

Reviewer 6: Does this mean I can have:
* Gateway in ns1, selecting route
* Route in ns2, targeting svc in ns3
* BackendTLSPolicy is ns3 targetting svc (no cross namespace allowed here), pointing to a cert in ns4
* Cert in ns4 with refgrant allowing backendTLS policy

So ultimately I have a chain of 4 namespaces allowing the Gateway is ns1 to read a secret in ns4?

Reviewer 4: That seems valid from what I can see, and is a good example of how quickly complexity can build  out of small changes.

Author: Yes, this seems to be a valid scenario, do you perceive this as a problem?

Reviewer 6: Maybe? The principles as I understand it is there should be a strong two way handshake between each cross namespace hop.

So the question we have is did the `ns4` owner authorization `ns1` to read its Secrets.

Working backwards:
* ns4 granted ns3 Secret read access explicitly with RefGrant. :heavy_check_mark: 
* ns3 explicitly granted ns2 access to **send to services**. :stop_sign: 
* ns1 explicitly granted ns2 access to bind routes :heavy_check_mark: 

I feel there is a violation here. There is no strong agreement from ns3 that ns2 can access its Secrets -- only that it can access its services.

Author: I agree with this concern - this calls at least for a clear warning when:
- BackendTLSPolicy is ns3 targetting svc (no cross namespace allowed here), pointing to a cert in ns4

ns3 owner should be aware that it is exposing the cert to all gateways that are allowed to **send to this service** will be able to access the certificate

Reviewer 6: I don't really think that is sufficient. Today, there may be many many users giving Service access. This is an **extremely low** privilege to grant, so users will hand it out freely.

Now we are saying that in a new release, that the persona is not even aware of, suddenly granting Service (low priv) gives access to Secret (high priv)

Author: How about requiring two grants, i.e. ?
- ns4 to ns3 secret (using secret to connect to the backend you control)
- ns1 to ns3 secret (to consume the secret)

Reviewer 5: This is a good point. A few options I can think of:

**1. References to Secrets that are used by the Gateway are relative to that Gateway**
With this option, we would consider the reference to this Secret as coming from the Gateway when it comes to cross-namespace logic. 

That means that a reference from BackendTLSPolicy in ns3 to a cert in ns1 would be considered _local_ and not need a ReferenceGrant, because the _Gateway_ is in ns1. On the other hand, a reference from BackendTLSPolicy in ns3 to a Secret in ns3 would require a ReferenceGrant that allowed a reference from the Gateway in ns1 to the secret in ns3.

**2. References are relative to BackendTLSPolicy, BackendTLSPolicy should be considered a privileged resource**

I've increasingly been making the case that Service owners aren't necessarily the correct owners for BackendTLSPolicy, particularly in larger organizations. If we consider BackendTLSPolicy a privileged resource, I think it's reasonable to say that valid references from this resource inherently grant access to the Gateway(s)  implementing that resource.

**3. Namespace-local for now**
I don't think we have clear evidence that we _need_ cross-namespace references at this level. Given the complexity involved, it may be safer to explicitly prohibit that as a starting point. We could instead require the Gateway, BackendTLSPolicy, and cert to all be in the same namespace when certs are specified at this level. That does admittedly feel rather onerous, but it's easier to add cross-namespace support in the future than to retroactively fix things if we leave things to open-ended to start with.

With all that said, I'm leaning towards the third option here. Open to other perspectives.

Author: I like the idea of requiring BackendTLSPolicy to be in the very same namespace as Gateway, but I am afraid this ship has sailed...  as per: https://gateway-api.sigs.k8s.io/api-types/backendtlspolicy/
"The BackendTLSPolicy and the Service must reside in the same namespace in order to prevent the complications involved with sharing trust across namespace boundaries."

Author: I believe we should park this discussion until we reach conclusion #3226, I am moving this overrides part to the future work.

  SubjectAltNames []SubjectAltName `json:"subjectAltNames,omitempty"`
}

// +kubebuilder:validation:Enum=Cookie;Header
type SubjectAltNameType string

const (
   // HostnameSubjectAltNameType specifies hostname-based SAN.
   //
   // Support: Core
   HostnameSubjectAltNameType SubjectAltNameType = "Hostname"

   // URISubjectAltNameType specifies URI-based SAN, e.g. SPIFFE id.
   //
   // Support: Core
   URISubjectAltNameType SubjectAltNameType = "URI"
)

type SubjectAltName struct {
  // Type determines the format of the Subject Alternative Name. Always required.
  Type SubjectAltNameType `json:"type"`

  // Hostname contains Subject Alternative Name specified in DNS name format. Required when Type is set to Hostname, ignored otherwise.
  Hostname v1.PreciseHostname `json:"hostname,omitempty"`

  // URI contains Subject Alternative Name specified in URI format. Required when Type is set to URI, ignored otherwise.
  URI string `json:"uri,omitempty"`
}
```

**2. Modify Spec for `BackendTLSPolicyValidation` `Hostname`**

Before:
```go
  // 2. Hostname MUST be used for authentication and MUST match the certificate
  //    served by the matching backend.
```

After:
```go

>   // 2. Only if SubjectAltNames is not specified, Hostname MUST be used for 
>   //    authentication and MUST match the certificate served by the matching
>   //    backend.

Reviewer 4: So, `Hostname` still sets the SNI that is sent as part of the initial request, but now no longer sets anything about how the server certificate must be checked?

In that case, I'd recommend something like this:
```suggestion
  // 2. Hostname MUST be used for authentication and MUST match the certificate
  //     served by the matching backend UNLESS SubjectAltNames is specified. If
  //     SubjectAltNames is specified, Hostname MUST only be used for setting
  //     the SNI sent, and the SubjectAltNames MUST be used for validating the
  //     serving certificate.
```

It's saying the same thing, but it is leaving less room for misunderstanding.

Author: Thanks a lot!

```

### Allow per-service mTLS settings BackendTLSPolicy

Gateway level TLS configuration already includes an `options` field. This has
been helpful for implementation-specific TLS configurations, or simply features
that have not made it to the core API yet. It would be similarly useful to have
an identical field on BackendTLSPolicy.

Examples:
- configuration options for vendor-specific mTLS automation

> - restrictions on the minimum supported TLS version or supported cipher suites

Reviewer 4: I just want to be on record as saying that I didn't like this field when we added it to Gateway, I don't like this field, but I don't have a good enough reason to block any changes. Let's see if it ends up misused in the wild.

Reviewer 5: I definitely agree with the hesitation here, but ultimately believe this is more helpful than not. There's such a wide range of TLS config possible that it seems unlikely that we'll ever cover all of the potential concepts in the core API. I think it's likely preferable to mirror the options map we have on Gateway Listeners than to leave out a form of extensibility here, though I'll admit that both options have risks.

  ClientCertificateRef SecretObjectReference `json:"clientCertificateRef,omitempty"`
}
```

#### Limitations

Configuring client certificate on the service level may result in multiple Gateways
from different vendors sharing the same identity when connecting to a single service.

However, given that:

1. per-service certificate configuration is a niche configuration, required to support
application-specific compatbility corner-cases
1. cross-namespace service usage is protected by the reference grants

this limitations should consitute an acceptable trade off.

The possible need for more fine-grained settings can be mitigated by allowing shared 
policies to selectively target specific Gateway instances, GatewayClasses or gateway 
namespaces.

### SANs on BackendTLSPolicy

This change enables the certificate to have a different identity than the SNI
(both are currently tied to the hostname field). This is particularly useful
when using SPIFFE, which relies on URI Subject Names which are not valid SNIs 
as per https://www.rfc-editor.org/rfc/rfc6066.html#section-3.

In such case either connection properties or an arbitrary SNI, like cluster-local 

> service name could be used for certificate selection, while the identity validation

Reviewer 5: Should hostname still be a required field and used for SNI? Or are you suggesting that it's optional and SNI can be _anything_ when SANs are specified?

Author: It cannot by anything, because it needs to be deterministic what we send to the backend. So yes hostname is still required.

   //
   // Support: Core
   URISubjectAltNameType SubjectAltNameType = "URI"
)

type SubjectAltName struct {
  // Type determines the format of the Subject Alternative Name. Always required.
  Type SubjectAltNameType `json:"type"`

  // Hostname contains Subject Alternative Name specified in DNS name format. Required when Type is set to Hostname, ignored otherwise.
  Hostname v1.PreciseHostname `json:"hostname,omitempty"`

  // URI contains Subject Alternative Name specified in URI format. Required when Type is set to URI, ignored otherwise.
  URI string `json:"uri,omitempty"`
}
```

**2. Modify Spec for `BackendTLSPolicyValidation` `Hostname`**

Before:
```go
  // 2. Hostname MUST be used for authentication and MUST match the certificate
  //    served by the matching backend.
```

After:
```go
  // 2. Hostname MUST be used for authentication and MUST match the certificate
  //     served by the matching backend UNLESS SubjectAltNames is specified. If
  //     SubjectAltNames is specified, Hostname MUST only be used for setting
  //     the SNI sent, and the SubjectAltNames MUST be used for validating the
  //     serving certificate.
```

> ### Allow per-service mTLS settings BackendTLSPolicy

Reviewer 5: ```suggestion
### Allow per-service TLS settings BackendTLSPolicy
```
